[JSC] Implement `Iterator.prototype.forEach` from Iterator Helpers Proposal

sosukesuzuki opened the pull request:
#### 0de8b68389b2ac948056557a94146997abbac0a0
<pre>
[JSC] Implement `Iterator.prototype.forEach` from Iterator Helpers Proposal
<a href="https://bugs.webkit.org/show_bug.cgi?id=279501">https://bugs.webkit.org/show_bug.cgi?id=279501</a>

Reviewed by Yusuke Suzuki.

This patch implements `Iterator.prototype.forEach`[1] from Iterator Helpers Proposal[2].

[1]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.foreach">https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.foreach</a>
[2]: <a href="https://github.com/tc39/proposal-iterator-helpers">https://github.com/tc39/proposal-iterator-helpers</a>

* JSTests/stress/iterator-prototype-forEach.js: Added.
(assert):
(sameArray):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameArray.Iter.prototype.get next):
(sameArray.Iter):
(sameArray.gen):
(const.validIter):
(const.invalidCallback.of.invalidCallbacks.shouldThrow):
(Iterator.prototype.forEach.call):
(const.iter.next):
(MyError):
(shouldThrow.const.validIter):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/283673@main">https://commits.webkit.org/283673@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05d7d7ed574ad197bfe9a2235899a8966c53a671

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67001 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71036 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18134 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17917 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53737 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12200 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70068 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42655 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57969 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34257 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39326 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15363 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16488 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60115 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61246 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15704 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72737 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66246 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15040 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61205 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10990 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58028 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61280 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14803 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9000 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2604 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88014 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42183 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15490 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43260 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44443 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43003 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
```diff
@@ -63,6 +67,9 @@ void JSIteratorPrototype::finishCreation(VM& vm, JSGlobalObject* globalObject)
         putDirectWithoutTransition(vm, vm.propertyNames->toArray, toArray, static_cast<unsigned>(PropertyAttribute::DontEnum));
+        // https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.foreach
+        JSFunction* forEach = JSFunction::create(vm, globalObject, 1, "forEach"_s, iteratorProtoFuncForEach, ImplementationVisibility::Public);
+        putDirectWithoutTransition(vm, vm.propertyNames->forEach, forEach, static_cast<unsigned>(PropertyAttribute::DontEnum));
```
Let's use `JSC_NATIVE_FUNCTION_WITHOUT_TRANSITION`

Constellation reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Implement `Iterator.prototype.forEach` from Iterator Helpers Proposal
https://bugs.webkit.org/show_bug.cgi?id=279501

Reviewed by Yusuke Suzuki.

This patch implements `Iterator.prototype.forEach`[1] from Iterator Helpers Proposal[2].

[1]: https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.foreach
[2]: https://github.com/tc39/proposal-iterator-helpers

* JSTests/stress/iterator-prototype-forEach.js: Added.
(assert):
(sameArray):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameArray.Iter.prototype.get next):
(sameArray.Iter):
(sameArray.gen):
(const.validIter):
(const.invalidCallback.of.invalidCallbacks.shouldThrow):
(Iterator.prototype.forEach.call):
(const.iter.next):
(MyError):
(shouldThrow.const.validIter):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/283673@main
